// vest/dist/es/classnames.development.js
import { invariant, hasOwnProperty, isPositive, isFunction } from 'vest-utils';
import { suiteSelectors } from 'vest';

// eslint-disable-next-line max-statements
function parse(summary) {
    invariant(summary && hasOwnProperty(summary, 'valid'), "Vest parser: expected argument at position 0 to be Vest's result object.");
    var sel = suiteSelectors(summary);
    var testedStorage = {};
    var selectors = {
        invalid: sel.hasErrors,
        tested: isTested,
        untested: isUntested,
        valid: sel.isValid,
        warning: sel.hasWarnings
    };
    return selectors;
    // Booleans
    function isTested(fieldName) {
        if (!fieldName) {
            return isPositive(summary.testCount);
        }
        if (hasOwnProperty(testedStorage, fieldName))
            return testedStorage[fieldName];
        testedStorage[fieldName] =
            hasOwnProperty(summary.tests, fieldName) &&
                isPositive(summary.tests[fieldName].testCount);
        return selectors.tested(fieldName);
    }
    function isUntested(fieldName) {
        return !(isPositive(summary.testCount) && selectors.tested(fieldName));
    }
}

/**
 * Creates a function that returns class names that match the validation result
 */
function classnames(res, classes) {
    if (classes === void 0) { classes = {}; }
    var selectors = parse(res);
    return function (key) {
        var classesArray = [];
        for (var selector in classes) {
            var sel = selector;
            if (isFunction(selectors[sel]) && selectors[sel](key)) {
                classesArray.push(classes[sel]);
            }
        }
        return classesArray.join(' ');
    };
}

export { classnames as default };
